File: hashTable/hash.ts
class HashValueable<Value> {
  [key: number]: Value;

  public add(value: Value) {
    const index = this.newIndex(value);
    if (this[index]) {
      throw `[${value}] Index ${index} already have a value: ${this[index]}`;
    }
    this[index] = value;
  }

  public getIndex(value: Value): number {
    const index = this.newIndex(value);
    if (!this[index]) {
      throw "No element found";
    }
    return index;
  }

  public remove(index: number) {
    if (!this[index]) {
      throw "No element to remove";
    }
    delete this[index];
  }

  public update(index: number, value: Value) {
    this.remove(index);
    this.add(value);
  }

  private toChars(value: string): string[] {
    return value.split("");
  }

  private hash(chars: string[]): number {
    let code: number = 0;
    for (let index = 0; index < chars.length; index++) {
      code += this.ascii(chars[index]) * (index + 1);
    }

    return code;
  }

  private newIndex(value: Value): number {
    let valueValueoString: string;
    if (typeof value === "object") {
      valueValueoString = JSON.stringify(value);
    } else {
      valueValueoString = String(value);
    }
    const chars = this.toChars(valueValueoString);
    const code = this.hash(chars);
    return code;
  }

  private ascii(char: string): number {
    return char.charCodeAt(0);
  }
}

export default HashValueable;

File: hashTable/hash.go
package main

import "fmt"

type element = map[int]string
type hashTable struct {
	items element
}

func NewHashTable() (ht hashTable) {
	ht.items = make(element)
	return
}

func (ht *hashTable) Add(key, value string) {
	i := hash(key)
	ht.items[i] = value
}

func (ht *hashTable) Remove(key string) {
	delete(ht.items, hash(key))
}

func (ht *hashTable) Get(key string) string {
	index := hash(key)
	return ht.items[index]
}

func hash(value string) (result int) {
	for index, char := range value {
		result = result + ascii(char)*index
	}
	return
}

func ascii(value rune) int {
	return int(byte(value))
}

func main() {
	example := NewHashTable()
	example.Add("foo", "bar")
	fmt.Println(example.Get("foo"))
}
